[IMP] account: days of sales outstanding per customer

hupo-odoo opened the pull request:
Add a read only field which enables to visualize the DSO ratio (days of sales oustanding) when determining a credit limit for a specific customer (on the partner page). This helps the user to know in how many days his customers pay their invoices.

task : 3196659




---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/113296).

fmdl commented on addons/account/models/partner.py:
```diff
@@ -677,3 +682,14 @@ def get_partner_localisation_fields_required_to_invoice(self, country_id):
         return []
+
+    def _compute_days_sales_outstanding(self):
+        for partner in self:
```
you use read_group
Warning amount_total is a field with currency, it is better to use amount_total_signed (computed in company currency).
And add ('company_id','=',self.env.company.id) in domain

and use fields.Date.conetxt_today(self), indeed fields.Date.today()

h4818 commented on addons/account/models/partner.py:
```diff
@@ -468,6 +468,11 @@ def _get_company_currency(self):
+        string='Days Sales Outstanding (DSO)',
+        help='[(Total Receivable/Total Revenue) * number of days since the first invoice] for this customer',
+        compute='_compute_days_sales_outstanding',
+        readonly=True)
```
I think computed fields are readonly by default

qdp-odoo commented on addons/account/models/partner.py:
```diff
@@ -565,6 +569,24 @@ def _compute_show_credit_limit(self):
+                [('id', 'child_of', partner.id)]).ids
+            first_invoice = self.env['account.move'].sudo().search(
+                domain=[('partner_id', 'in', all_partner_ids),
+                        ('move_type', '=', 'out_invoice'),
```
```suggestion
                        ('move_type', 'in', ('out_invoice', 'out_receipt')),
```

qdp-odoo replied:
you also probably need to add a clause on the current company, btw :thinking:

qdp-odoo commented on addons/account/models/partner.py:
```diff
@@ -565,6 +569,24 @@ def _compute_show_credit_limit(self):
+        for partner in self:
+            all_partner_ids = self.with_context(active_test=False).search(
+                [('id', 'child_of', partner.id)]).ids
+            first_invoice = self.env['account.move'].sudo().search(
```
why do you need to call sudo()?

qdp-odoo commented on addons/account/models/partner.py:
```diff
@@ -565,6 +569,24 @@ def _compute_show_credit_limit(self):
+                        ('move_type', '=', 'out_invoice'),
+                        ('state', 'not in', ['cancel', 'draft'])],
+                order='invoice_date asc',
+                limit=1)
```
doing a search inside of a loop is not efficient, you'd go ahead of performance issues.

qdp-odoo reviewed (changes requested):
need some rework/explanation :)

h4818 commented on addons/account/models/partner.py:
```diff
@@ -565,6 +569,28 @@ def _compute_show_credit_limit(self):
+                partner.days_sales_outstanding = 0
+                continue
+
+            oldest_invoice = min(combined_invoices, key=lambda x: x.invoice_date)
```
It seems like `combined_invoices` is used to find the oldest invoice.
this seems to duplicate the logic in `_invoice_total`...
will it not be better if they both shared the same function `_partner_invoices_domain`: `_invoice_totals` will use it for a `read group`, whereas you will use it with `limit=1, order="date DESC"`

(I'm not sure it will work considering `_invoice_total` uses `account.invoice.report` - might be worth a try though - to prevent discrepancies in the future)
perhaps @qdp-odoo has an opinion on this

hupo-odoo commented:
Hello @qdp-odoo first of all thank you for the review you have made on this PR. I have rebased in order to avoid conflict errors. Would you mind have a look at the corrected code when you have time ? Thank you

Levizar commented on addons/account/models/partner.py:
```diff
@@ -406,6 +406,29 @@ def _invoice_total(self):
+            for commercial_partner, oldest_invoice_date, total_invoiced_tax_included in self.env['account.move']._read_group(
+                domain=[
+                    ('state', 'not in', ['draft', 'cancel']),
+                    ('move_type', 'in', ('out_invoice', 'out_refund', 'out_receipt')),
```
```suggestion
                    ('move_type', 'in', self.env["account.move"].get_sale_types(include_receipts=True)),
```

Levizar reviewed (changes requested):
Nice job! We're almost there.

Could you also add a test to solidify the feature on the long term?
That would ensure the feature doesn't get wiped out during a huge refactoring/revamping/whatever

hupo-odoo commented:
Thank you @Levizar for your last review. I have added a test and also changed the function from `_read_group()` to `read_group()`. It needed some little adaptation but I think it's the function I am supposed to use

Levizar reviewed (approved):
@robodoo r+
